## Setup

In [1]:
import os
from pathlib import Path
import sys

# If we're using Google Colab, we set the environment variable to point to the relevant folder in our Google Drive:
if 'COLAB_GPU' in os.environ:
    from google.colab import drive
    drive.mount('/content/drive')
    os.environ['SKIN_LESION_CLASSIFICATION'] = '/content/drive/MyDrive/Colab Notebooks/skin-lesion-classification'

# Otherwise, we use the environment variable on our local system:
project_environment_variable = "SKIN_LESION_CLASSIFICATION"

# Path to the root directory of the project:
project_path = Path(os.environ.get(project_environment_variable))

# Relative path to /scripts (from where custom modules will be imported):
scripts_path = project_path.joinpath("scripts")

# Add this path to sys.path so that Python will look there for modules:
sys.path.append(str(scripts_path))

# Now import path_step from our custom utils module to create a dictionary to all subdirectories in our root directory:
from utils import path_setup
path = path_setup.subfolders(project_path)

path['project'] : D:\projects\skin-lesion-classification
path['images'] : D:\projects\skin-lesion-classification\images
path['models'] : D:\projects\skin-lesion-classification\models
path['expository'] : D:\projects\skin-lesion-classification\expository
path['literature'] : D:\projects\skin-lesion-classification\literature
path['notebooks'] : D:\projects\skin-lesion-classification\notebooks
path['presentation'] : D:\projects\skin-lesion-classification\presentation
path['scripts'] : D:\projects\skin-lesion-classification\scripts


## Loading and preprocessing metadata

In [2]:
from typing import Type, Union      # For type hints
from processing import process      # Custom module for processing metadata

data_dir: Path = path["images"]     # Path to directory containing metadata.csv file
csv_filename: str = "metadata.csv"  # The filename
    
restrict_to: Union[dict, None] = None                   # Remove all records *unless* column k lies in list v, for k : v in restrict_to dictionary.    
remove_if: Union[dict, None] = None                     # Remove all records if column k lies in list v, for k : v in remove_if dictionary.    
drop_row_if_missing_value_in: Union[list, None] = None  # We drop all rows for which there is a missing value in a column from this list.   
                                    
tvr: int = 3              # Ratio of training set to validation set. See discussion below for explanation.
seed: int = 0             # Random seed for parts of the process where randomness is called for.
keep_first: bool = False  # If False, then, for each lesion, we choose a random image to assign to our training set. 
stratified: bool = True   # If True, we stratify classes so that the proportions remain as stable as possible after train/val split. 
                          # If False, the proportions will be roughly similar.

to_classify: Union[list, dict] = ["mel",   # These are the lesion types we are interested in classifying. 
                                  "bcc",   # Any missing ones will be grouped together as the 0-label class: no need to write "other" here.
                                  "akiec", # If 'other' is not desired, use restrict_to attribute above
                                  "nv",]

train_one_img_per_lesion: Union[bool, None] = False # If False, we take advantage of the (in some cases) multiple images of a lesion in our dataset
val_one_img_per_lesion: Union[bool, None] = False   # If False, we will validate our model by combining multiple predictions for a lesion (for multiple images of it) into a single prediction
val_expansion_factor: Union[int, None] = 3          # A random transformation may be applied to an image before making a prediction.
                                                    # For a given lesion, we may make multiple predictions (as specified here), and combine them into a single prediction.
    
sample_size: Union[None, dict] = {"mel": 2000,     # Handling class imbalance by upsampling minority classes/downsampling majority classes     
                                  "bcc": 2000,     # Specify how many images of each lesion diagnosis we want in our training set.
                                  "akiec": 2000, 
                                  "nv": 2000,
                                  "other" : 2000,} # Could also leave out "other" here, and include e.g. "df: 2000" if we wanted to.    

In [3]:
# Create an instance of the process class with attribute values as above.
demo = process(data_dir=data_dir,
               csv_filename=csv_filename,
               restrict_to=restrict_to,
               remove_if=remove_if,
               drop_row_if_missing_value_in=drop_row_if_missing_value_in,
               tvr=tvr,
               seed=seed,
               keep_first=keep_first,
               stratified=stratified,
               to_classify=to_classify,
               train_one_img_per_lesion=train_one_img_per_lesion,
               val_expansion_factor=val_expansion_factor,
               sample_size=sample_size,)

- Loaded file 'D:\projects\skin-lesion-classification\images\metadata.csv'.
- Inserted 'num_images' column in dataframe, to the right of 'lesion_id' column.
- Inserted 'label' column in dataframe, to the right of 'dx' column: 
  {'bkl': 0, 'df': 0, 'vasc': 0, 'akiec': 1, 'nv': 2, 'mel': 3, 'bcc': 4}
- Added 'set' column to dataframe, with values 't1', 'v1', 'ta', and 'va', to the right of 'localization' column.
- Basic, overall dataframe (pre-train/test split): self.df
- Balancing classes in training set.
- Balanced training set (uses as many different images per lesion as possible): self.df_train
- Expanding validation set: will combine 3 predictions into one, for each lesion in val set.
- Expanded validation set (one image per lesion, repeated 3 times): self.df_val1
- Expanded validation set (use up to 3 different images per lesion, if available): self.df_val_a
- Small sample dataframes for code testing: self._df_train_code_test, self._df_val1_code_test, self._df_val_a_code_test


## Class distribution

In [4]:
for across in ["lesions", "images"]:
    for subset in ["all", "train", "val"]:
        process.dx_dist(demo, subset = subset, across = across)


DISTRIBUTION OF LESIONS BY DIAGNOSIS: OVERALL



dx         nv   other     mel     bcc   akiec
freq  5403.00  898.00  614.00  327.00  228.00
%       72.33   12.02    8.22    4.38    3.05

Total lesions: 7470.


DISTRIBUTION OF LESIONS BY DIAGNOSIS: TRAIN



dx         nv   other     mel     bcc   akiec
freq  4052.00  673.00  460.00  245.00  171.00
%       72.34   12.02    8.21    4.37    3.05

Total lesions: 5601 (74.98% of all lesions).


DISTRIBUTION OF LESIONS BY DIAGNOSIS: VAL



dx         nv   other     mel    bcc  akiec
freq  1351.00  225.00  154.00  82.00  57.00
%       72.28   12.04    8.24   4.39   3.05

Total lesions: 1869 (25.02% of all lesions).


DISTRIBUTION OF IMAGES BY DIAGNOSIS: OVERALL



dx         nv    other      mel     bcc   akiec
freq  6705.00  1356.00  1113.00  514.00  327.00
%       66.95    13.54    11.11    5.13    3.27

Total images: 10015.


DISTRIBUTION OF IMAGES BY DIAGNOSIS: TRAIN



dx         nv    other     mel     bcc   akiec
freq  5007.00  1008.00  831.00  384.00  250.00
%       66.94    13.48   11.11    5.13    3.34

Total images: 7480 (74.69% of all images).


DISTRIBUTION OF IMAGES BY DIAGNOSIS: VAL



dx         nv   other     mel     bcc  akiec
freq  1698.00  348.00  282.00  130.00  77.00
%       66.98   13.73   11.12    5.13   3.04

Total images: 2535 (25.31% of all images).



## Train/val split

<!-- <details>
    <summary><b><i>Train test split explanation: click here to expand/collapse</i></b></summary> -->
    
We partition our dataset based on ```lesion_id```, **not** on ```image_id```: that way, every lesion will be represented in training or in validation, but not both.

For each classification task, we will train a model by making use of
* **exactly one** image for every lesion in our training set;
* **all** images of every lesion in our training set.

In both cases, we will vaildate our model by making use of 
* **exactly one** image for every lesion in our validation set;
* **all** images of every lesion in our validation set (at least, _potentially_ all of them). 

**However**, we will make only one prediction per lesion (```lesion_id```) in our validation set: if there are multiple images of a lesion in the validation set, we will combine the predictions for the multiple images into a single prediction for the lesion.

Accordingly, we proceed as follows. We'll explain by example, assuming the dataset is not filtered before splitting (if it is, the number of distinct lesions will be less than $7470$, and the proportions will be different).
1. Randomly select (without replacement) a proportion of our $7470$ distinct ```lesion_id```s and label them with ```t``` (train). 
2. Label the remaining ```lesion_id```s with ```v``` (validate).
3. For each ```lesion_id``` labeled with a ```t```:
    * Select an ```image_id``` and label it ```t1```.
    * Label all (if any) remaining ```image_id```s corresponding to this ```lesion_id``` with ```ta```.
4.  For each ```lesion_id``` labeled with a ```v```:
    * Select an ```image_id``` and label it ```v1```.
    * Label all (if any) remaining ```image_id```s corresponding to this ```lesion_id``` with ```va```.

In Step 1, the number of ```lesion_id```s randomly selected to be labeled ```t``` will be such that the ratio of ```t```s to ```v```s is as close as possible to a specified ratio ```tvr``` (we default to $3$, i.e. $\approx75\%$ of lesions are represented in training). In Steps 3 and 4, the first substep can be done randomly (our default choice), or we can simply choose the "first" image in our table that corresponds to the lesion (see ```keep_first``` attribute of the ```process``` class). 

The four train/val scenarios we could consider are:
* ```t1v1```: train on precisely those images labeled ```t1``` and validate on precisely those labeled ```v1```.
* ```t1va```: train on precisely those images labeled ```t1``` and validate on precisely those labeled ```v1``` **or** ```va```.
* ```tav1```: train on precisely those images labeled ```t1``` **or** ```ta``` and validate on precisely those labeled ```v1```.
* ```tava```: train on precisely those images labeled ```t1``` **or** ```ta``` and validate on precisely those labeled ```v1``` ***or*** ```va```.

The mnemonic is ```t``` for training, ```v``` for validation, ```1``` for one-image-per-lesion, and ```a``` for all images.
<!-- </details> -->

In [5]:
# Let's have a look at our metadata dataframe, which is now just an attribute of the metadata instance of the process class.
from utils import print_header

instance = demo
df = instance.df

print_header("First five rows of metadata table")

to_print = ["Added columns\n".upper(), 
            "\'num_images\': number of images of lesion in dataset", 
            "\'label\': class to which lesion belongs",
            "\'set\': train/val assignment",
            "\'t*\': lesion is in the training set",
            "\'v*\': lesion is in the validation set",
            "\'t1\': we would train on this image if training a model on exactly one, or on all, image(s) per lesion in the training set",
            "If training set is balanced using one image per lesion, this one image would be re-used as many times as necessary.",
            "\'ta\': we would train on this image if training a model on all images of each lesion in the training set",
            "If training set is balanced using all images per lesion, images labeled ta would all be used before any image is repeated.",
            "\'v1': we\'d use this image if validating a model on exactly one, or on all, image(s) per lesion in the validation set",
            "If a validation expansion factor is given, this one image would be re-used that many times",
            "\'va': we\'d use this image if validating on all images of each lesion in the validation set" ,
            "If a validation expansion factor is given, iamges labeled va would all be used before any image is repeated.",
            "NB: if more than one image is used for any lesion in validation, the predictions will be combined into a single prediction"
           ]

print("\n- ".join(to_print))
display(df.head())


FIRST FIVE ROWS OF METADATA TABLE

ADDED COLUMNS

- 'num_images': number of images of lesion in dataset
- 'label': class to which lesion belongs
- 'set': train/val assignment
- 't*': lesion is in the training set
- 'v*': lesion is in the validation set
- 't1': we would train on this image if training a model on exactly one, or on all, image(s) per lesion in the training set
- If training set is balanced using one image per lesion, this one image would be re-used as many times as necessary.
- 'ta': we would train on this image if training a model on all images of each lesion in the training set
- If training set is balanced using all images per lesion, images labeled ta would all be used before any image is repeated.
- 'v1': we'd use this image if validating a model on exactly one, or on all, image(s) per lesion in the validation set
- If a validation expansion factor is given, this one image would be re-used that many times
- 'va': we'd use this image if validating on all images of ea

lesion_id  num_images      image_id   dx  label dx_type   age   sex  \
0  HAM_0000118           2  ISIC_0027419  bkl      0   histo  80.0  male   
1  HAM_0000118           2  ISIC_0025030  bkl      0   histo  80.0  male   
2  HAM_0002730           2  ISIC_0026769  bkl      0   histo  80.0  male   
3  HAM_0002730           2  ISIC_0025661  bkl      0   histo  80.0  male   
4  HAM_0001466           2  ISIC_0031633  bkl      0   histo  75.0  male   

  localization set  
0        scalp  ta  
1        scalp  t1  
2        scalp  va  
3        scalp  v1  
4          ear  va

## Balancing the training set

<!-- <details>
    <summary><b><i>Balancing/upsampling explanation: click here to expand/collapse</i></b></summary> -->

We explain the balancing procedure by way of example. (This is performed by the ```balance``` method of the ```process``` class in our ```processing``` module.) We assume the dataset has not been filtered, training to validation ratio is $3$, etc. There are $460$ distinct melanoma lesions represented in our training set. As most melanoma are represented by multiple distinct images, there are a total of $831$ distinct images of melanoma lesions in our training set. Suppose we want our training set to contain $2000$ melanoma images: each of the $460$ distinct melanoma lesions will be represented by $2000/460 \approx 4.35$ images on average. We do not merely sample with replacement.

The goal is to (a) have as little variance as possible in the number of times a lesion is represented, and (b) use as many distinct images as possible (taking advantage of the fact that there are multiple _distinct_ images of most melanoma). Thus, we note that $2000 = 4\times 460 + 160$, so we will use each of the $460$ distinct melanoma lesions four times, and make the remainder up by randomly sample $160$ distinct lesions from the $160$. In other words, exactly $300$ distinct lesions will each be represented by exactly four images, and exactly $160$ distinct lesions will each be represented by exactly five images: $2000 = 300 \times 4 + 160 \times 5$. 

How do we select the four images of each distinct melanoma lesion (plus another one image for $160$ of them)? Consider lesion id ```HAM_0000871``` for example: there are three distinct images of this lesion in our data set. Thus, if ```train_one_img_per_lesion``` is ```False```, we will use all three of them, and then randomly select one more (or two more if this particular lesion were to be one of the $160$ that are represented five times). See below. On the other hand, if ```train_one_img_per_lesion``` is ```True```, we have no choice but to use the one image (label ```t1```) four times.
    
<!-- </details> -->

In [6]:
# from utils import print_header

# The specific numbers in this example assume a certain choice for the attributes, including 
# sample_size: Union[None, dict] = {"mel": 2000,         
#                                   "bcc": 2000, 
#                                   "akiec": 2000, 
#                                   "nv": 2000,
#                                   "other" : 2000,}

instance = demo
df = demo.df_train

print_header("Eg: Representations of lesion HAM_0000871 in balanced training set")

to_print = ["HAM_0000871 represented by four images\n".upper(),
            "Three distinct images of this lesion to choose from: ISIC_0025964, ISIC_0030623, and ISIC_0025964",
            "Use ISIC_0025964 once, ISIC_0030623 twice, and ISIC_0025964 once",]

print("\n- ".join(to_print))

display(df[df['lesion_id'] == 'HAM_0000871'])


EG: REPRESENTATIONS OF LESION HAM_0000871 IN BALANCED TRAINING SET

HAM_0000871 REPRESENTED BY FOUR IMAGES

- Three distinct images of this lesion to choose from: ISIC_0025964, ISIC_0030623, and ISIC_0025964
- Use ISIC_0025964 once, ISIC_0030623 twice, and ISIC_0025964 once


lesion_id  lesion_mult  num_images      image_id  img_mult   dx  \
1773  HAM_0000871            4           3  ISIC_0025964         2  mel   
1774  HAM_0000871            4           3  ISIC_0025964         2  mel   
1775  HAM_0000871            4           3  ISIC_0030623         1  mel   
3087  HAM_0000871            4           3  ISIC_0026506         1  mel   

      label dx_type   age     sex localization set  
1773      3   histo  40.0  female        chest  ta  
1774      3   histo  40.0  female        chest  ta  
1775      3   histo  40.0  female        chest  t1  
3087      3   histo  40.0  female        trunk  ta

In [7]:
# from utils import print_header
# The specific numbers given in this example assume a certain choice for the attributes, including 
# sample_size: Union[None, dict] = {"mel": 2000,         
#                                   "bcc": 2000, 
#                                   "akiec": 2000, 
#                                   "nv": 2000,
#                                   "other" : 2000,}

instance = demo
df = demo.df_train
df = df[df['set'].isin(["ta", "t1"]) & (df['dx'] == 'mel')]

print_header("Eg: Melanoma in balanced training set")

to_print = ["Value counts for \'lesion_mult\' column\n".upper(),
            "300 distinct melanoma lesions each represented by four images: 300*4 = 1200",
            "160 distinct melanoma lesions each represented by five images: 160*5 = 800",]

print("\n- ".join(to_print[:3]))
display(df['lesion_mult'].value_counts())

print("\n- ".join(to_print[3:]))

display(df)


EG: MELANOMA IN BALANCED TRAINING SET

VALUE COUNTS FOR 'LESION_MULT' COLUMN

- 300 distinct melanoma lesions each represented by four images: 300*4 = 1200
- 160 distinct melanoma lesions each represented by five images: 160*5 = 800


4    1200
5     800
Name: lesion_mult, dtype: int64

lesion_id  lesion_mult  num_images      image_id  img_mult   dx  \
1773  HAM_0000871            4           3  ISIC_0025964         1  mel   
1774  HAM_0000871            4           3  ISIC_0030623         1  mel   
1775  HAM_0000040            5           1  ISIC_0027190         5  mel   
1776  HAM_0000040            5           1  ISIC_0027190         5  mel   
1777  HAM_0000040            5           1  ISIC_0027190         5  mel   
...           ...          ...         ...           ...       ...  ...   
7773  HAM_0002552            5           3  ISIC_0032936         2  mel   
7781  HAM_0002552            5           3  ISIC_0033232         2  mel   
7782  HAM_0002552            5           3  ISIC_0033232         2  mel   
9998  HAM_0003521            5           2  ISIC_0032258         2  mel   
9999  HAM_0003521            5           2  ISIC_0032258         2  mel   

      label dx_type   age     sex     localization set  
1773      1   histo  40.0  female            chest  ta  
1774      1   histo  40.0  female            chest  t1  
1775      1   histo  80.0    male  upper extremity  t1  
1776      1   histo  80.0    male  upper extremity  t1  
1777      1   histo  80.0    male  upper extremity  t1  
...     ...     ...   ...     ...              ...  ..  
7773      1   histo  25.0    male  upper extremity  ta  
7781      1   histo  25.0    male  upper extremity  ta  
7782      1   histo  25.0    male  upper extremity  ta  
9998      1   histo  70.0  female             back  ta  
9999      1   histo  70.0  female             back  ta  

[2000 rows x 12 columns]

## Expanding the validation set

As mentioned already, we make one prediction per lesion. However, we may have multiple images of a given lesion at our disposal: we could make a prediction for each of them and combine them somehow into a single prediction for the lesion. Even if there is only one image of a lesion, we could make multiple predictions on it: if a random transformation is applied to an image before our model makes a prediction on it, this would yield a different array of probabilities each time. Again, we could combine the results into a single prediction.

This is what the attribute ```val_expansion_factor``` of the ```process``` class is concerned with. Similarly to the way we balance the training set, we can replicate one single image per lesion in the validation set as many times as specified by ```val_expansion_factor```, as in ```self.df_val1```, or we can take advantage of other images of the lesion (if available), as in ```self.val_a```.

In [7]:
# from utils import print_header
# The specific numbers given in this example assume a certain choice for the attributes  

instance = demo

df = demo.df_val1
df = df[df['set'].isin(["va", "v1"]) & (df['dx'] == 'mel')]

print_header("Eg: Melanoma in expanded validation set (only one image per lesion used)")

to_print = [f"- Note that \'lesion_mult\' is always {instance.val_expansion_factor}",
            "HAM_0005678 represented by three images",
            "Two distinct images of this lesion: ISIC_0031023 and ISIC_0028086",
            f"However, only use ISIC_0031023 ({instance.val_expansion_factor} times)",]

print("\n- ".join(to_print))

display(df)

df = demo.df_val_a
df = df[df['set'].isin(["va", "v1"]) & (df['dx'] == 'mel')]

print_header("Eg: Melanoma in expanded validation set (all images used)")

to_print = [f"- Note that \'lesion_mult\' is always {instance.val_expansion_factor}",
            "HAM_0005678 represented by three images",
            "Two distinct images of this lesion to choose from: ISIC_0031023 and ISIC_0028086",
            "Use ISIC_0031023 once, and ISIC_0028086 twice",]

print("\n- ".join(to_print))

display(df)


EG: MELANOMA IN EXPANDED VALIDATION SET (ONLY ONE IMAGE PER LESION USED)

- Note that 'lesion_mult' is always 3
- HAM_0005678 represented by three images
- Two distinct images of this lesion: ISIC_0031023 and ISIC_0028086
- However, only use ISIC_0031023 (3 times)


lesion_id  lesion_mult  num_images      image_id  img_mult   dx  \
603   HAM_0005678            3           2  ISIC_0031023         3  mel   
604   HAM_0005678            3           2  ISIC_0031023         3  mel   
605   HAM_0005678            3           2  ISIC_0031023         3  mel   
606   HAM_0006722            3           2  ISIC_0031499         3  mel   
607   HAM_0006722            3           2  ISIC_0031499         3  mel   
...           ...          ...         ...           ...       ...  ...   
1060  HAM_0004081            3           1  ISIC_0031957         3  mel   
1061  HAM_0004081            3           1  ISIC_0031957         3  mel   
1062  HAM_0004746            3           2  ISIC_0028764         3  mel   
1063  HAM_0004746            3           2  ISIC_0028764         3  mel   
1064  HAM_0004746            3           2  ISIC_0028764         3  mel   

      label dx_type   age     sex     localization set  
603       3   histo  60.0    male            chest  v1  
604       3   histo  60.0    male            chest  v1  
605       3   histo  60.0    male            chest  v1  
606       3   histo  85.0  female  lower extremity  v1  
607       3   histo  85.0  female  lower extremity  v1  
...     ...     ...   ...     ...              ...  ..  
1060      3   histo  70.0  female  lower extremity  v1  
1061      3   histo  70.0  female  lower extremity  v1  
1062      3   histo  65.0  female             back  v1  
1063      3   histo  65.0  female             back  v1  
1064      3   histo  65.0  female             back  v1  

[462 rows x 12 columns]


EG: MELANOMA IN EXPANDED VALIDATION SET (ALL IMAGES USED)

- Note that 'lesion_mult' is always 3
- HAM_0005678 represented by three images
- Two distinct images of this lesion to choose from: ISIC_0031023 and ISIC_0028086
- Use ISIC_0031023 once, and ISIC_0028086 twice


lesion_id  lesion_mult  num_images      image_id  img_mult   dx  \
603   HAM_0005678            3           2  ISIC_0031023         1  mel   
604   HAM_0005678            3           2  ISIC_0028086         2  mel   
605   HAM_0005678            3           2  ISIC_0028086         2  mel   
606   HAM_0006722            3           2  ISIC_0030443         1  mel   
607   HAM_0006722            3           2  ISIC_0031499         2  mel   
...           ...          ...         ...           ...       ...  ...   
1060  HAM_0004746            3           2  ISIC_0028764         2  mel   
1061  HAM_0004746            3           2  ISIC_0028764         2  mel   
1062  HAM_0004746            3           2  ISIC_0029021         1  mel   
1063  HAM_0002525            3           2  ISIC_0025188         1  mel   
1064  HAM_0001953            3           2  ISIC_0025611         1  mel   

      label dx_type   age     sex     localization set  
603       3   histo  60.0    male            chest  v1  
604       3   histo  60.0    male            chest  va  
605       3   histo  60.0    male            chest  va  
606       3   histo  85.0  female  lower extremity  va  
607       3   histo  85.0  female  lower extremity  v1  
...     ...     ...   ...     ...              ...  ..  
1060      3   histo  65.0  female             back  v1  
1061      3   histo  65.0  female             back  v1  
1062      3   histo  65.0  female             back  va  
1063      3   histo  55.0    male             face  va  
1064      3   histo  65.0    male             back  va  

[462 rows x 12 columns]

## Fine-tuning EfficientNet or ResNet18

### Transforms

In [8]:
# Now let's set values for the attributes of our resnet18 class (the model we will use with out processed data).
# One of the attributes has to do with image transformations.

import torchvision.transforms as transforms

transform = transforms.Compose([
    
transforms.RandomCrop((300, 300)),
transforms.Resize((224,224)), # Resize images to fit ResNet input size
transforms.ToTensor(),
transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),  # Normalize with ImageNet stats
])    

In [9]:
import pandas as pd
from typing import Union, List, Callable
import torchvision.models as models

source: Union[process, pd.DataFrame] = demo      # Processed data to be fed into model for training.
                                                 # Must either be an instance of the process class, or a dataframe of the same format as source.df if source were an instance of the process class.
model_dir: Path = path["models"]                 # Path to directory where models/model info/model results are stored.
transform: Union[None, 
                 transforms.Compose, 
                 List[Callable]] = transform     # Transform to be applied to images before feeding into neural network.
batch_size: Union[None, int] = 32                # Mini-batch size: default 32.
epochs: Union[None, int] = 10                    # Number of epochs (all layers unfrozen from the start): default 10.
base_learning_rate: Union[None, float] = 1/1000  # Learning rate to start with: default 1/1000. Using Adam optimizer.
filename_stem: Union[None, str] = "rn18"         # For saving model and related files. Default "rn18" (if ResNet model) or "EffNet" (if EfficientNet), or "cnn".
filename_suffix: Union[None, str] = "demo"       # Something descriptive and unique for future reference. Default empty string "".
overwrite: Union[None, bool] = True              # If False, any will generate an unused filename for saving .pth, .csv files etc., but appending a two-digit number.
                                                 # If None, will default to False. Only set to True if confident that training done on previous instances with same filename stem and suffix can be over-written.
code_test: Union[None, bool] = True
# model: Union[None, models.ResNet, models.EfficientNet] = models.efficientnet_b0(weights=models.EfficientNet_B0_Weights.DEFAULT) # Pre-trained model. Default: ResNet18.   
model: Union[None, models.ResNet, models.EfficientNet] = models.resnet18(weights="ResNet18_Weights.DEFAULT")  

In [10]:
# Create an instance of the resnet18 class with attribute values as above.
from multiclass_models import cnn

resnet_demo = cnn(                                   
    source=source,                                           
    model_dir=model_dir,
    transform=transform,
    batch_size=batch_size,
    epochs=epochs,                                          
    base_learning_rate=base_learning_rate,
    filename_stem=filename_stem,
    filename_suffix=filename_suffix,                         
    overwrite=overwrite,
    code_test=code_test,    
    model=model
)


CODE TEST MODE

- self.epochs set to 1
- self.Print set to True
- self.filename_suffix set to 'test'
- self.overwrite set to True
- self.df_train, self.df_val1, self.df_val_a replaced with a small number of records
- Change code_test attribute to False and re-create/create new cnn instance after testing is done.

Existing files will be overwritten. 
Base filename: rn18_ta_bal_test_1e_test_00
Attributes saved to file: D:\projects\skin-lesion-classification\models\rn18_ta_bal_test_1e_test_00_attributes.json


### Code test with small batch of images

In [11]:
# from utils import print_header

instance = resnet_demo

print_header("Code test: training set")
print(f"{instance.df_train.shape[0]} images".upper())
display(instance.df_train.head())

print_header(f"Code test: validation set (one image per lesion, repeated {instance.source.val_expansion_factor} times)")
print(f"{instance.df_val1.shape[0]} images".upper())
display(instance.df_val1.head())

print_header(f"Code test: validation set ({instance.source.val_expansion_factor} possibly different images per lesion)")
print(f"{instance.df_val_a.shape[0]} images".upper())
display(instance.df_val_a.head())


CODE TEST: TRAINING SET

170 IMAGES


lesion_id  lesion_mult  num_images      image_id  img_mult   dx  label  \
0  HAM_0003768            3           2  ISIC_0029929         1  bkl      0   
1  HAM_0004928            3           2  ISIC_0031424         1  bkl      0   
2  HAM_0004928            3           2  ISIC_0029770         2  bkl      0   
3  HAM_0004928            3           2  ISIC_0029770         2  bkl      0   
4  HAM_0003768            3           2  ISIC_0026634         2  bkl      0   

  dx_type   age   sex     localization set  
0   histo  80.0  male  upper extremity  t1  
1   histo  65.0  male             neck  t1  
2   histo  65.0  male             neck  ta  
3   histo  65.0  male             neck  ta  
4   histo  80.0  male  upper extremity  ta


CODE TEST: VALIDATION SET (ONE IMAGE PER LESION, REPEATED 3 TIMES)

42 IMAGES


lesion_id  lesion_mult  num_images      image_id  img_mult   dx  label  \
0  HAM_0003218            3           1  ISIC_0033305         3  bkl      0   
1  HAM_0003218            3           1  ISIC_0033305         3  bkl      0   
2  HAM_0003218            3           1  ISIC_0033305         3  bkl      0   
3  HAM_0000983            3           1  ISIC_0033490         3  bkl      0   
4  HAM_0000983            3           1  ISIC_0033490         3  bkl      0   

     dx_type   age      sex localization set  
0  consensus  75.0     male         back  v1  
1  consensus  75.0     male         back  v1  
2  consensus  75.0     male         back  v1  
3  consensus   NaN  unknown      unknown  v1  
4  consensus   NaN  unknown      unknown  v1


CODE TEST: VALIDATION SET (3 POSSIBLY DIFFERENT IMAGES PER LESION)

42 IMAGES


lesion_id  lesion_mult  num_images      image_id  img_mult   dx  label  \
0  HAM_0004406            3           2  ISIC_0034125         1  bkl      0   
1  HAM_0004406            3           2  ISIC_0033060         2  bkl      0   
2  HAM_0004406            3           2  ISIC_0033060         2  bkl      0   
3  HAM_0003943            3           3  ISIC_0031078         1  bkl      0   
4  HAM_0003943            3           3  ISIC_0031464         1  bkl      0   

  dx_type   age     sex     localization set  
0   histo  80.0    male             back  va  
1   histo  80.0    male             back  v1  
2   histo  80.0    male             back  v1  
3   histo  80.0  female  lower extremity  va  
4   histo  80.0  female  lower extremity  v1

In [12]:
# Train the model on the specified training data by calling the train method:
# from utils import print_header

instance = resnet_demo

print_header("Code test: training and validation")
instance.train()


CODE TEST: TRAINING AND VALIDATION

image_id, label, ohe-label: ISIC_0031831, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0029891, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0029713, 1, tensor([0., 1., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0025350, 1, tensor([0., 1., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0033278, 3, tensor([0., 0., 0., 1., 0.])
image_id, label, ohe-label: ISIC_0033065, 2, tensor([0., 0., 1., 0., 0.])
image_id, label, ohe-label: ISIC_0025927, 3, tensor([0., 0., 0., 1., 0.])
image_id, label, ohe-label: ISIC_0029278, 4, tensor([0., 0., 0., 0., 1.])
image_id, label, ohe-label: ISIC_0029713, 1, tensor([0., 1., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0030344, 1, tensor([0., 1., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0028314, 1, tensor([0., 1., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0025350, 1, tensor([0., 1., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0033860, 0, tensor([1., 0., 0., 0., 0.])
i

image_id, label, ohe-label: ISIC_0031272, 4, tensor([0., 0., 0., 0., 1.])
image_id, label, ohe-label: ISIC_0024932, 3, tensor([0., 0., 0., 1., 0.])
image_id, label, ohe-label: ISIC_0025599, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0032652, 4, tensor([0., 0., 0., 0., 1.])
image_id, label, ohe-label: ISIC_0025196, 1, tensor([0., 1., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0024997, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0025628, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0029099, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0030344, 1, tensor([0., 1., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0031669, 2, tensor([0., 0., 1., 0., 0.])
image_id, label, ohe-label: ISIC_0025927, 3, tensor([0., 0., 0., 1., 0.])
image_id, label, ohe-label: ISIC_0029929, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0032114, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_00302

image_id, label, ohe-label: ISIC_0026313, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0026629, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0026629, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0026629, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0030443, 3, tensor([0., 0., 0., 1., 0.])
image_id, label, ohe-label: ISIC_0031499, 3, tensor([0., 0., 0., 1., 0.])
image_id, label, ohe-label: ISIC_0031499, 3, tensor([0., 0., 0., 1., 0.])
image_id, label, ohe-label: ISIC_0033299, 3, tensor([0., 0., 0., 1., 0.])
image_id, label, ohe-label: ISIC_0033299, 3, tensor([0., 0., 0., 1., 0.])
image_id, label, ohe-label: ISIC_0033299, 3, tensor([0., 0., 0., 1., 0.])
image_id, label, ohe-label: ISIC_0032692, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0033608, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_0033969, 0, tensor([1., 0., 0., 0., 0.])
image_id, label, ohe-label: ISIC_00286

In [13]:
# from utils import print_header
from multiclass_models import load_dict

# Let's look at the training and validation loss for each epoch:
instance = resnet_demo

print_header("Code test")
print(f"Loss dictionary (TRAINING AND VALIDATION LOSS FROM EACH EPOCH)".upper())

try:
    if instance.epoch_losses is not None:
        display(instance.epoch_losses)
    else:
        retrieved_epoch_losses = load_dict(instance.model_dir, instance._filename + "_epoch_losses")
        display(retrieved_epoch_losses)
except:
    retrieved_epoch_losses = load_dict(instance.model_dir, instance._filename + "_epoch_losses")
    display(retrieved_epoch_losses)


CODE TEST

LOSS DICTIONARY (TRAINING AND VALIDATION LOSS FROM EACH EPOCH)


{'train_loss': array([0.71549305]),
 'val1_loss': array([5.0177567]),
 'val_a_loss': array([3.87107873])}

In [14]:
# from utils import print_header
import torch
import torch.nn as nn

instance = resnet_demo

print_header("Code test: model architecture and state dictionary")

if instance.state_dict is None:
    print("Loading model and state dictionary from file\n".upper())
    file_path_pth = instance.model_dir.joinpath(instance._filename + ".pth")

    # model = models.efficientnet_b0()  
    model = models.resnet18()  
    if isinstance(model,models.ResNet):
        num_ftrs = model.fc.in_features
        model.fc = nn.Linear(num_ftrs, len(instance.label_codes))
    elif isinstance(model,models.EfficientNet):
        num_ftrs = model.classifier[1].in_features
        model.classifier[1] = nn.Linear(num_ftrs, len(instance.label_codes))

    # Load the state dictionary into the model
    state_dict = torch.load(file_path_pth)
    model.load_state_dict(state_dict)

    instance.model = model
    instance.state_dict = state_dict
    
print_header("Code test:  model architecture")
print(f"Note: \'out_features={len(instance.label_codes)}\' at the end".upper())
display(instance.model)


CODE TEST: MODEL ARCHITECTURE AND STATE DICTIONARY


CODE TEST:  MODEL ARCHITECTURE

NOTE: 'OUT_FEATURES=5' AT THE END


ResNet(
  (conv1): Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
  (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (relu): ReLU(inplace=True)
  (maxpool): MaxPool2d(kernel_size=3, stride=2, padding=1, dilation=1, ceil_mode=False)
  (layer1): Sequential(
    (0): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (1): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
  

In [15]:
print_header("Code test: model state dictionary")
print(str(instance.state_dict)[:1000], "\n ... LOTS OF PARAMETERS ...\n", str(instance.state_dict)[-1000:])


CODE TEST: MODEL STATE DICTIONARY

OrderedDict([('conv1.weight', tensor([[[[-1.2006e-02, -5.9770e-03, -1.8481e-03,  ...,  5.5908e-02,
            1.6073e-02, -1.4027e-02],
          [ 9.5711e-03,  9.5411e-03, -1.1017e-01,  ..., -2.7173e-01,
           -1.2985e-01,  2.4616e-03],
          [-8.6098e-03,  5.9050e-02,  2.9509e-01,  ...,  5.1917e-01,
            2.5565e-01,  6.2269e-02],
          ...,
          [-2.8976e-02,  1.5986e-02,  7.2161e-02,  ..., -3.3399e-01,
           -4.2163e-01, -2.5939e-01],
          [ 2.8915e-02,  4.0837e-02,  6.2291e-02,  ...,  4.1259e-01,
            3.9243e-01,  1.6449e-01],
          [-1.5208e-02, -3.4787e-03, -2.4442e-02,  ..., -1.5184e-01,
           -8.3447e-02, -7.1686e-03]],

         [[-1.0895e-02, -2.5281e-02, -3.3266e-02,  ...,  3.1528e-02,
           -3.0424e-03, -2.9929e-02],
          [ 4.6378e-02,  3.5028e-02, -1.0307e-01,  ..., -3.1316e-01,
           -1.6438e-01, -5.5776e-03],
          [-3.2989e-03,  9.9680e-02,  4.0311e-01,  ...,  7.06

In [16]:
# from utils import print_header
from multiclass_models import get_probabilities

instance = resnet_demo

# model = models.efficientnet_b0()  
# model = models.resnet18() 
# if isinstance(model,models.ResNet):
#     num_ftrs = model.fc.in_features
#     model.fc = nn.Linear(num_ftrs, len(instance.label_codes))
# elif isinstance(model,models.EfficientNet):
#     num_ftrs = model.classifier[1].in_features
#     model.classifier[1] = nn.Linear(num_ftrs, len(instance.label_codes))

instance.df_probabilities_val1 = get_probabilities(df=instance.df_val1,
                                                   data_dir=instance.data_dir,
                                                   model_dir=instance.model_dir,
                                                   model=instance.model,
                                                   filename=instance._filename,
                                                   label_codes=instance.label_codes,
                                                   transform=instance.transform,
                                                   batch_size=instance.batch_size,
                                                   Print=False,
                                                   save_as=instance._filename,)

instance.df_probabilities_val_a = get_probabilities(df=instance.df_val_a,
                                                    data_dir=instance.data_dir,
                                                    model_dir=instance.model_dir,
                                                    model=instance.model,
                                                    filename=instance._filename,
                                                    label_codes=instance.label_codes,
                                                    transform=instance.transform,
                                                    batch_size=instance.batch_size,
                                                    Print=False,
                                                    save_as=instance._filename,)

print_header("Code test: probabilities, validation set (one image per lesion)")
display_columns = ['lesion_id', 'image_id', 'dx'] + [col for col in instance.df_probabilities_val1.columns if col.startswith('prob')]
display(instance.df_probabilities_val1[display_columns].head())

print_header("Code test: probabilities, validation set (more than one image per lesion)")
display(instance.df_probabilities_val_a[display_columns].head())

Saving probabilities: D:\projects\skin-lesion-classification\models\rn18_ta_bal_test_1e_test_00_probabilities.csv
Saving probabilities: D:\projects\skin-lesion-classification\models\rn18_ta_bal_test_1e_test_00_probabilities.csv

CODE TEST: PROBABILITIES, VALIDATION SET (ONE IMAGE PER LESION)



lesion_id      image_id   dx    prob_other  prob_akiec   prob_nv  \
0  HAM_0003218  ISIC_0033305  bkl  7.460868e-04    0.001333  0.000606   
1  HAM_0003218  ISIC_0033305  bkl  1.372970e-04    0.000913  0.000118   
2  HAM_0003218  ISIC_0033305  bkl  3.386477e-03    0.005549  0.006373   
3  HAM_0000983  ISIC_0033490  bkl  2.949595e-07    0.000010  0.000002   
4  HAM_0000983  ISIC_0033490  bkl  2.715284e-06    0.000006  0.000003   

   prob_mel  prob_bcc  
0  0.122082  0.875233  
1  0.139352  0.859480  
2  0.278105  0.706586  
3  0.000018  0.999971  
4  0.000057  0.999931


CODE TEST: PROBABILITIES, VALIDATION SET (MORE THAN ONE IMAGE PER LESION)



lesion_id      image_id   dx  prob_other  prob_akiec   prob_nv  prob_mel  \
0  HAM_0004406  ISIC_0034125  bkl    0.170367    0.002337  0.186573  0.075787   
1  HAM_0004406  ISIC_0033060  bkl    0.000015    0.000763  0.000012  0.010129   
2  HAM_0004406  ISIC_0033060  bkl    0.000021    0.000561  0.000011  0.003380   
3  HAM_0003943  ISIC_0031078  bkl    0.000010    0.000002  0.000005  0.987654   
4  HAM_0003943  ISIC_0031464  bkl    0.000020    0.000007  0.000058  0.978615   

   prob_bcc  
0  0.564936  
1  0.989082  
2  0.996028  
3  0.012329  
4  0.021300

In [17]:
# from utils import print_header
from multiclass_models import df_from_ids, get_probabilities     

instance = resnet_demo
df = instance.df

print_header("Code test: prediction on individual images or lesions")

to_print = ["- We can make predictions for individual images or lesions.",
            "We only require a dataframe with an \'image_id\' column.",
            "Given the filename of an image, the df_from_ids function will construct such a dataframe.",
            "We can then feed this dataframe into the get_probabilities function.",
            "Here is the result of passing \'filenames=[\'ISIC_0033305\',\'ISIC_0025661\']\' to df_from_ids:",
            "- And here are the corresponding probabilities:",
            "- If we have a lesion_id with associated image_ids, we can also construct a small dataframe.",
            "The df_from_ids function also takes arguments for the number of predictions we want to make for a given image/lesion.",
            "- Here is the result of passing \'lesion_ids=\'HAM_0000118'\', \'multiplicity=3\', and \'one_img_per_lesion=False\' to df_from_ids:",
            "However, here we have filtered all columns except for lesion_id and image_id (knowing the diagnosis defeats the purpose).",
            "- Here are the associated probabilities:",
            "- Notice that the probabilities may vary with each execution of a prediction.",
            "This is because a random transformation may be applied to each image before our model makes a prediction on it."]

df_2img = df_from_ids(filenames=['ISIC_0033305','ISIC_0025661'], # can be a string or a list of strings
                       multiplicity=None,
                       lesion_ids=None,
                       df=df,
                       one_img_per_lesion=None,)

print("\n- ".join(to_print[:5]))

display(df_2img)

df_2img_prob = get_probabilities(df=df_2img,
                  data_dir=instance.data_dir,
                  model_dir=instance.model_dir,
                  model=instance.model,
                  filename=instance._filename,
                  label_codes=instance.label_codes,
                  transform=instance.transform,
                  batch_size=instance.batch_size,
                  Print=False,
                  save_as=None,)   

print(to_print[5])
display(df_2img_prob)

print("\n- ".join(to_print[6:8]))

df_1les = df_from_ids(filenames=None,
                       multiplicity=3,
                       lesion_ids='HAM_0000118', # can be a string or a list of strings
                       df=df,
                       one_img_per_lesion=False,)

print("\n- ".join(to_print[8:10]))
display_columns = ['lesion_id', 'image_id'] 
display(df_1les[display_columns])

df_1les_prob = get_probabilities(df=df_1les,
                  data_dir=instance.data_dir,
                  model_dir=instance.model_dir,
                  model=instance.model,
                  filename=instance._filename,
                  label_codes=instance.label_codes,
                  transform=instance.transform,
                  batch_size=instance.batch_size,
                  Print=False,
                  save_as=None,)   

print(to_print[10])
display_columns = ['lesion_id', 'image_id'] + [col for col in df_1les_prob if col.startswith('prob')]
display(df_1les_prob[display_columns])

print("\n- ".join(to_print[11:]))


CODE TEST: PREDICTION ON INDIVIDUAL IMAGES OR LESIONS

- We can make predictions for individual images or lesions.
- We only require a dataframe with an 'image_id' column.
- Given the filename of an image, the df_from_ids function will construct such a dataframe.
- We can then feed this dataframe into the get_probabilities function.
- Here is the result of passing 'filenames=['ISIC_0033305','ISIC_0025661']' to df_from_ids:


image_id
0  ISIC_0033305
1  ISIC_0025661

- And here are the corresponding probabilities:


image_id    prob_other  prob_akiec       prob_nv  prob_mel  prob_bcc
0  ISIC_0033305  1.133565e-03    0.002940  1.143219e-03  0.083207  0.911576
1  ISIC_0025661  3.843240e-08    0.000205  7.986385e-08  0.000002  0.999793

- If we have a lesion_id with associated image_ids, we can also construct a small dataframe.
- The df_from_ids function also takes arguments for the number of predictions we want to make for a given image/lesion.
- Here is the result of passing 'lesion_ids='HAM_0000118'', 'multiplicity=3', and 'one_img_per_lesion=False' to df_from_ids:
- However, here we have filtered all columns except for lesion_id and image_id (knowing the diagnosis defeats the purpose).


lesion_id      image_id
0  HAM_0000118  ISIC_0027419
1  HAM_0000118  ISIC_0027419
2  HAM_0000118  ISIC_0025030

- Here are the associated probabilities:


lesion_id      image_id  prob_other  prob_akiec   prob_nv  prob_mel  \
0  HAM_0000118  ISIC_0027419    0.003104    0.055993  0.002405  0.472147   
1  HAM_0000118  ISIC_0027419    0.002975    0.026197  0.003501  0.696560   
2  HAM_0000118  ISIC_0025030    0.000016    0.008122  0.000022  0.004272   

   prob_bcc  
0  0.466352  
1  0.270767  
2  0.987568

- Notice that the probabilities may vary with each execution of a prediction.
- This is because a random transformation may be applied to each image before our model makes a prediction on it.


In [219]:
import importlib

import multiclass_models

importlib.reload(multiclass_models)

<module 'multiclass_models' from 'D:\\projects\\skin-lesion-classification\\scripts\\multiclass_models.py'>

In [81]:
from collections import OrderedDict
from multiclass_models import threshold, get_argmax, append_prediction, df_with_probabilities, mode_with_random, predictions_mode

instance = resnet18mc_test

# try:
#     instance._inference_df = df_with_probabilities(instance)
# except:
filename = instance._filename + "_infer.csv" # or whatever it might be
file_path_csv = instance.model_dir.joinpath(filename)
file_path_csv
instance._inference_df = df_with_probabilities(file_path_csv)

print_header("Code test: probabilities for all images".upper())

display(instance._inference_df)

threshold_dict_help = OrderedDict([('mel',0.4)])
threshold_dict_hinder = OrderedDict([('nv',0.6)])

inverse_label_codes = {value: key for key, value in instance.label_codes.items()}

print_header("Code test: probabilities and predictions for all images".upper())
instance._predictions_df = append_prediction(original_df=instance._inference_df, 
                                             probabilities_df=instance._inference_df, 
                                             threshold_dict_help=threshold_dict_help,
                                             threshold_dict_hinder=threshold_dict_hinder,
                                             inverse_label_codes=inverse_label_codes,)

display(instance._predictions_df)

print_header("Code test: probabilities and predictions for images in validation set".upper())
instance._predictions_df_val = instance._predictions_df[instance._predictions_df["set"] == "v1"]

display(instance._predictions_df_val)

print_header("Code test: probabilities and aggregated [...] predictions for images in validation set".upper())
display(predictions_mode(instance._predictions_df))


CODE TEST: PROBABILITIES FOR ALL IMAGES



lesion_id  num_images      image_id     dx  label dx_type   age  \
0    HAM_0000118           2  ISIC_0027419    bkl      0   histo  80.0   
1    HAM_0000118           2  ISIC_0025030    bkl      0   histo  80.0   
2    HAM_0006663           2  ISIC_0033482    bkl      0   histo  85.0   
3    HAM_0006663           2  ISIC_0033201    bkl      0   histo  85.0   
4    HAM_0003162           2  ISIC_0032997    bkl      0   histo  75.0   
..           ...         ...           ...    ...    ...     ...   ...   
139  HAM_0000967           1  ISIC_0025539  akiec      4   histo  75.0   
140  HAM_0005480           2  ISIC_0026149  akiec      4   histo  65.0   
141  HAM_0005480           2  ISIC_0025992  akiec      4   histo  65.0   
142  HAM_0004592           2  ISIC_0030730  akiec      4   histo  60.0   
143  HAM_0004592           2  ISIC_0029141  akiec      4   histo  60.0   

        sex     localization set  prob_other  prob_mel  prob_bcc   prob_nv  \
0      male            scalp  ta    0.002810  0.488154  0.320958  0.112851   
1      male            scalp  t1    0.006403  0.546590  0.207624  0.160286   
2    female  lower extremity  t1    0.001759  0.126734  0.534769  0.037566   
3    female  lower extremity  ta    0.005804  0.150670  0.624709  0.059166   
4      male             back  t1    0.058898  0.161691  0.408894  0.197394   
..      ...              ...  ..         ...       ...       ...       ...   
139    male  lower extremity  v1    0.109885  0.118041  0.390637  0.202687   
140    male             face  va    0.013788  0.298523  0.205440  0.246306   
141    male             face  v1    0.001394  0.167860  0.157719  0.106153   
142  female             face  v1    0.006594  0.468394  0.284184  0.187424   
143  female             face  va    0.000264  0.174940  0.325974  0.258017   

     prob_akiec  
0      0.075228  
1      0.079096  
2      0.299172  
3      0.159651  
4      0.173124  
..          ...  
139    0.178750  
140    0.235942  
141    0.566874  
142    0.053403  
143    0.240804  

[144 rows x 15 columns]


CODE TEST: PROBABILITIES AND PREDICTIONS FOR ALL IMAGES



lesion_id  num_images      image_id     dx  label dx_type   age  \
0    HAM_0000118           2  ISIC_0027419    bkl      0   histo  80.0   
1    HAM_0000118           2  ISIC_0025030    bkl      0   histo  80.0   
2    HAM_0006663           2  ISIC_0033482    bkl      0   histo  85.0   
3    HAM_0006663           2  ISIC_0033201    bkl      0   histo  85.0   
4    HAM_0003162           2  ISIC_0032997    bkl      0   histo  75.0   
..           ...         ...           ...    ...    ...     ...   ...   
139  HAM_0000967           1  ISIC_0025539  akiec      4   histo  75.0   
140  HAM_0005480           2  ISIC_0026149  akiec      4   histo  65.0   
141  HAM_0005480           2  ISIC_0025992  akiec      4   histo  65.0   
142  HAM_0004592           2  ISIC_0030730  akiec      4   histo  60.0   
143  HAM_0004592           2  ISIC_0029141  akiec      4   histo  60.0   

        sex     localization set  prob_other  prob_mel  prob_bcc   prob_nv  \
0      male            scalp  ta    0.002810  0.488154  0.320958  0.112851   
1      male            scalp  t1    0.006403  0.546590  0.207624  0.160286   
2    female  lower extremity  t1    0.001759  0.126734  0.534769  0.037566   
3    female  lower extremity  ta    0.005804  0.150670  0.624709  0.059166   
4      male             back  t1    0.058898  0.161691  0.408894  0.197394   
..      ...              ...  ..         ...       ...       ...       ...   
139    male  lower extremity  v1    0.109885  0.118041  0.390637  0.202687   
140    male             face  va    0.013788  0.298523  0.205440  0.246306   
141    male             face  v1    0.001394  0.167860  0.157719  0.106153   
142  female             face  v1    0.006594  0.468394  0.284184  0.187424   
143  female             face  va    0.000264  0.174940  0.325974  0.258017   

     prob_akiec  pred  
0      0.075228     1  
1      0.079096     1  
2      0.299172     2  
3      0.159651     2  
4      0.173124     2  
..          ...   ...  
139    0.178750     2  
140    0.235942     1  
141    0.566874     4  
142    0.053403     1  
143    0.240804     2  

[144 rows x 16 columns]


CODE TEST: PROBABILITIES AND PREDICTIONS FOR IMAGES IN VALIDATION SET



lesion_id  num_images      image_id     dx  label    dx_type   age  \
109  HAM_0006801           3  ISIC_0032570    bkl      0      histo  60.0   
110  HAM_0002335           2  ISIC_0031812    bkl      0      histo  85.0   
112  HAM_0006988           1  ISIC_0024943    bkl      0  consensus  75.0   
113  HAM_0005586           1  ISIC_0024415     nv      3  follow_up  45.0   
115  HAM_0001076           2  ISIC_0033627     nv      3      histo  65.0   
116  HAM_0006855           1  ISIC_0030842     nv      3      histo  50.0   
118  HAM_0004330           2  ISIC_0030555     df      0      histo  70.0   
120  HAM_0007418           2  ISIC_0026313     df      0  consensus  50.0   
121  HAM_0006371           3  ISIC_0034135     df      0  consensus  35.0   
124  HAM_0004126           2  ISIC_0030060    mel      1      histo  80.0   
127  HAM_0000042           3  ISIC_0033905    mel      1      histo  45.0   
129  HAM_0001113           1  ISIC_0031900    mel      1      histo  55.0   
130  HAM_0001719           1  ISIC_0031090   vasc      0      histo  55.0   
131  HAM_0004877           2  ISIC_0027159   vasc      0      histo  85.0   
133  HAM_0005245           1  ISIC_0024375   vasc      0  consensus  70.0   
135  HAM_0000393           2  ISIC_0027865    bcc      2      histo  40.0   
136  HAM_0003851           1  ISIC_0031643    bcc      2      histo  70.0   
138  HAM_0004940           2  ISIC_0026970    bcc      2      histo  80.0   
139  HAM_0000967           1  ISIC_0025539  akiec      4      histo  75.0   
141  HAM_0005480           2  ISIC_0025992  akiec      4      histo  65.0   
142  HAM_0004592           2  ISIC_0030730  akiec      4      histo  60.0   

        sex     localization set  prob_other  prob_mel  prob_bcc   prob_nv  \
109  female  lower extremity  v1    0.046393  0.083066  0.102322  0.734325   
110    male  upper extremity  v1    0.043069  0.150440  0.688703  0.109991   
112  female             face  v1    0.000012  0.003398  0.116960  0.000692   
113  female             back  v1    0.102728  0.366121  0.024346  0.502185   
115    male             back  v1    0.423836  0.146423  0.157589  0.152320   
116    male             back  v1    0.003835  0.209552  0.454396  0.091031   
118    male  lower extremity  v1    0.435236  0.063799  0.146160  0.350733   
120    male  lower extremity  v1    0.014638  0.335821  0.276991  0.305636   
121  female  lower extremity  v1    0.018797  0.126853  0.207847  0.636889   
124  female             back  v1    0.236065  0.137967  0.447473  0.084833   
127  female  upper extremity  v1    0.004573  0.164309  0.332060  0.182941   
129    male             back  v1    0.000392  0.091109  0.227575  0.016306   
130  female  upper extremity  v1    0.526077  0.095319  0.188934  0.155346   
131  female  upper extremity  v1    0.869906  0.031418  0.053604  0.031890   
133  female          abdomen  v1    0.137608  0.200343  0.217249  0.430879   
135  female          abdomen  v1    0.002556  0.256503  0.485825  0.231674   
136    male             back  v1    0.000886  0.077194  0.844768  0.052705   
138    male             back  v1    0.026466  0.239001  0.300750  0.414656   
139    male  lower extremity  v1    0.109885  0.118041  0.390637  0.202687   
141    male             face  v1    0.001394  0.167860  0.157719  0.106153   
142  female             face  v1    0.006594  0.468394  0.284184  0.187424   

     prob_akiec  pred  
109    0.033894     3  
110    0.007796     2  
112    0.878939     4  
113    0.004621     1  
115    0.119832     0  
116    0.241185     2  
118    0.004073     0  
120    0.066914     1  
121    0.009614     3  
124    0.093662     2  
127    0.316116     2  
129    0.664617     4  
130    0.034324     0  
131    0.013182     0  
133    0.013921     2  
135    0.023442     2  
136    0.024447     2  
138    0.019127     2  
139    0.178750     2  
141    0.566874     4  
142    0.053403     1


CODE TEST: PROBABILITIES AND AGGREGATED [...] PREDICTIONS FOR IMAGES IN VALIDATION SET



lesion_id  num_images      image_id     dx  label dx_type   age  \
0    HAM_0000118           2  ISIC_0027419    bkl      0   histo  80.0   
1    HAM_0000118           2  ISIC_0025030    bkl      0   histo  80.0   
2    HAM_0006663           2  ISIC_0033482    bkl      0   histo  85.0   
3    HAM_0006663           2  ISIC_0033201    bkl      0   histo  85.0   
4    HAM_0003162           2  ISIC_0032997    bkl      0   histo  75.0   
..           ...         ...           ...    ...    ...     ...   ...   
139  HAM_0000967           1  ISIC_0025539  akiec      4   histo  75.0   
140  HAM_0005480           2  ISIC_0026149  akiec      4   histo  65.0   
141  HAM_0005480           2  ISIC_0025992  akiec      4   histo  65.0   
142  HAM_0004592           2  ISIC_0030730  akiec      4   histo  60.0   
143  HAM_0004592           2  ISIC_0029141  akiec      4   histo  60.0   

        sex     localization set  prob_other  prob_mel  prob_bcc   prob_nv  \
0      male            scalp  ta    0.002810  0.488154  0.320958  0.112851   
1      male            scalp  t1    0.006403  0.546590  0.207624  0.160286   
2    female  lower extremity  t1    0.001759  0.126734  0.534769  0.037566   
3    female  lower extremity  ta    0.005804  0.150670  0.624709  0.059166   
4      male             back  t1    0.058898  0.161691  0.408894  0.197394   
..      ...              ...  ..         ...       ...       ...       ...   
139    male  lower extremity  v1    0.109885  0.118041  0.390637  0.202687   
140    male             face  va    0.013788  0.298523  0.205440  0.246306   
141    male             face  v1    0.001394  0.167860  0.157719  0.106153   
142  female             face  v1    0.006594  0.468394  0.284184  0.187424   
143  female             face  va    0.000264  0.174940  0.325974  0.258017   

     prob_akiec  pred_mode  
0      0.075228          1  
1      0.079096          1  
2      0.299172          2  
3      0.159651          2  
4      0.173124          2  
..          ...        ...  
139    0.178750          2  
140    0.235942          1  
141    0.566874          1  
142    0.053403          1  
143    0.240804          1  

[144 rows x 16 columns]

In [77]:
predictions_mode(instance._predictions_df)

lesion_id  num_images      image_id     dx  label dx_type   age  \
0    HAM_0000118           2  ISIC_0027419    bkl      0   histo  80.0   
1    HAM_0000118           2  ISIC_0025030    bkl      0   histo  80.0   
2    HAM_0006663           2  ISIC_0033482    bkl      0   histo  85.0   
3    HAM_0006663           2  ISIC_0033201    bkl      0   histo  85.0   
4    HAM_0003162           2  ISIC_0032997    bkl      0   histo  75.0   
..           ...         ...           ...    ...    ...     ...   ...   
139  HAM_0000967           1  ISIC_0025539  akiec      4   histo  75.0   
140  HAM_0005480           2  ISIC_0026149  akiec      4   histo  65.0   
141  HAM_0005480           2  ISIC_0025992  akiec      4   histo  65.0   
142  HAM_0004592           2  ISIC_0030730  akiec      4   histo  60.0   
143  HAM_0004592           2  ISIC_0029141  akiec      4   histo  60.0   

        sex     localization set  prob_other  prob_mel  prob_bcc   prob_nv  \
0      male            scalp  ta    0.002810  0.488154  0.320958  0.112851   
1      male            scalp  t1    0.006403  0.546590  0.207624  0.160286   
2    female  lower extremity  t1    0.001759  0.126734  0.534769  0.037566   
3    female  lower extremity  ta    0.005804  0.150670  0.624709  0.059166   
4      male             back  t1    0.058898  0.161691  0.408894  0.197394   
..      ...              ...  ..         ...       ...       ...       ...   
139    male  lower extremity  v1    0.109885  0.118041  0.390637  0.202687   
140    male             face  va    0.013788  0.298523  0.205440  0.246306   
141    male             face  v1    0.001394  0.167860  0.157719  0.106153   
142  female             face  v1    0.006594  0.468394  0.284184  0.187424   
143  female             face  va    0.000264  0.174940  0.325974  0.258017   

     prob_akiec  pred_mode  
0      0.075228          1  
1      0.079096          1  
2      0.299172          2  
3      0.159651          2  
4      0.173124          2  
..          ...        ...  
139    0.178750          2  
140    0.235942          1  
141    0.566874          1  
142    0.053403          1  
143    0.240804          1  

[144 rows x 16 columns]

In [79]:
from evaluation import weighted_average_f, confusion_matrix_with_metric

instance = resnet18mc_test
map_labels = instance.label_codes

A = instance._predictions_df_val['label']
B = instance._predictions_df_val['pred']
AxB = pd.crosstab(A,B,margins=True,dropna=False)

beta = 1
weights = None

instance._cm = confusion_matrix_with_metric(AxB=AxB,
                                            lst=None,
                                            full_pad=True,
                                            func=weighted_average_f,
                                            beta=beta,
                                            weights=weights,
                                            percentage=False,
                                            map_labels=map_labels)

print_header("Code test: confusion matrix (validation set)")

to_print = ["- The overall evaluation metric would appear at the bottom right, if it were defined (this code test set is too small).",
            "It would be a class-wise weighted average fbeta score, beta and weights as specified (default values 1).",
            "One could also pass None, a float, or a function other than weighted_average_f to the func parameter in confusion_matrix_with_metric."]
print("\n- ".join(to_print))
display(instance._cm.fillna('_'))


CODE TEST: CONFUSION MATRIX (VALIDATION SET)

- The overall evaluation metric would appear at the bottom right, if it were defined (this code test set is too small).
- It would be a class-wise weighted average fbeta score, beta and weights as specified (default values 1).
- One could also pass None, a float, or a function other than weighted_average_f to the func parameter in confusion_matrix_with_metric.


predicted other  mel       bcc   nv     akiec All    recall
actual                                                     
other         3    1         2    2         1   9  0.333333
mel           0    0         2    0         1   3       0.0
bcc           0    0         3    0         0   3       1.0
nv            1    1         1    0         0   3       0.0
akiec         0    1         1    0         1   3  0.333333
All           4    3         9    2         3  21         _
precision  0.75  0.0  0.333333  0.0  0.333333   _         _

In [80]:
from evaluation import metric_dictionary
# import numpy as np
# import pandas as pd

instance = resnet18mc_test

target = instance._predictions_df['label'] 
prediction = instance._predictions_df['pred'] 
probabilities = instance._predictions_df.filter(regex=r'^prob_')

# target = pd.concat([target_, pd.Series([1,2])])
# target = target.reset_index(drop=True)
# prediction = pd.concat([prediction_, pd.Series([0,2])]) 
# prediction = prediction.reset_index(drop=True)
# probabilities = probabilities_.copy()
# probabilities.loc[16] = 0.2
# probabilities.loc[17] = 0.2
# probabilities = probabilities.values

print_header("Code test: other metrics")
to_print = ["- ACC: accuracy",
            "BACC: balanced accuracy",
            "precision: macro-averaged precision (equal weight to each class)",
            "recal: macro-averaged recall (equal weight to each class)",
            "Fbeta: macro-averaged F_beta score (equal weight to each class)",
            "MCC: Matthews correlation coefficient",
            "ROC-AUC mac: macro-averaged ROC-AUC (equal weight to each class)",
            "ROC-AUC wt: weighted-average ROC-AUC (larger class -> more weight)",
            "ROC-AUC wt*: weighted-average ROC-AUC (larger class -> *less weight)",            
            ]

instance._metric_dict = metric_dictionary(target=target, 
                                          prediction=prediction, 
                                          probabilities=probabilities)
print("\n- ".join(to_print))
display(pd.DataFrame(instance._metric_dict))


CODE TEST: OTHER METRICS

- ACC: accuracy
- BACC: balanced accuracy
- precision: macro-averaged precision (equal weight to each class)
- recal: macro-averaged recall (equal weight to each class)
- Fbeta: macro-averaged F_beta score (equal weight to each class)
- MCC: Matthews correlation coefficient
- ROC-AUC mac: macro-averaged ROC-AUC (equal weight to each class)
- ROC-AUC wt: weighted-average ROC-AUC (larger class -> more weight)
- ROC-AUC wt*: weighted-average ROC-AUC (larger class -> *less weight)


ACC      BACC  precision    recall      F1/2        F1        F2  \
0  0.277778  0.350072   0.346472  0.350072  0.283516  0.271494  0.298296   

        MCC  ROC-AUC mac  ROC-AUC wt  ROC-AUC wt*  
0  0.158465     0.722044     0.72177     0.726919

### Baseline model

In [22]:
# import torchvision.transforms as transforms

transform = transforms.Compose([
    
transforms.RandomCrop((300, 300)),
transforms.Resize((224,224)), # Resize images to fit ResNet input size
transforms.ToTensor(),
transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),  # Normalize with ImageNet stats
])    

In [23]:
# import pandas as pd
# from typing import List, Callable

df: pd.DataFrame = metadata.df                   # Background dataset for the model. metadata._df_sample_batch is a random selection of 64 rows of metadata.df. We use it for testing our code.
train_set: Union[pd.DataFrame, list, str] = "t1" # "t1" (one image per lesion in training set); ["t1", "ta"] (all images for each lesion in training set); can also specify another sub-dataframe of self.df.
val_set: Union[pd.DataFrame, list, str] = "v1"   # Similar to train_set above.
label_codes: dict = metadata._label_codes        # Correspondence between label codes like 0 and label words like 'other'.
data_dir: Path = path["images"]                  # Path to directory where images are stored.
model_dir: Path = path["models"]                 # Path to directory where models/model info/model results are stored.
transform: List[Callable] = transform            # Transform to be applied to images before feeding to ResNet-18
batch_size: int = 32                             # Mini-batch size: default 32.
epochs: int = 10                                 # Number of epochs (all layers unfrozen from the start): default 10.
base_learning_rate: float = 0.001                # Learning rate to start with: default 0.001. Using Adam optimizer.
filename_stem: str = "rn18mc"                    # For saving model and related files. train set and num epochs will be appended automatically. Default "rn18mc".
filename_suffix: str = "base"                    # Something descriptive and unique for future reference and to avoid over-writing other files. Default empty string "".

In [24]:
resnet18mc_base = resnet18(                                  
    df=df, 
    train_set=train_set,
    val_set=val_set,
    label_codes=label_codes,
    data_dir=data_dir,
    model_dir=model_dir,
    transform=transform,
    batch_size=batch_size,
    epochs=epochs,                                                
    base_learning_rate=base_learning_rate,
    filename_stem=filename_stem,
    filename_suffix=filename_suffix,                                  
)

In [ ]:
# We only need to do this once...
# resnet18mc_base.train()

In [404]:
import importlib

del instance._inference_df
import evaluation

importlib.reload(evaluation)

<module 'evaluation' from 'D:\\projects\\skin-lesion-classification\\scripts\\evaluation.py'>

In [413]:
# from evaluation import get_argmax, append_prediction, get_probabilities_df

instance = resnet18mc_base

# try:
#     instance._inference_df = df_with_probabilities(instance)
# except:
filename = "rn18mc_t1_10e_base_inference.csv" #instance._filename + "_infer.csv" # "rn18mc_t1_10e_base_inference.csv"
file_path_csv = instance.model_dir.joinpath(filename)
instance._inference_df = df_with_probabilities(file_path_csv)
instance._inference_df = pd.merge(instance.df, instance._inference_df, on='image_id', how='inner')

print_header("Baseline model: probabilities for all images".upper())

display(instance._inference_df.head())

inverse_label_codes = {value: key for key, value in instance.label_codes.items()}

print_header("Baseline model: probabilities and predictions for all images".upper())
instance._predictions_df = append_prediction(original_df=instance._inference_df, 
                                        probabilities_df=instance._inference_df, 
                                        inverse_label_codes=inverse_label_codes,)

display(instance._predictions_df.head())

print_header("Baseline model: probabilities and predictions for images in validation set".upper())
instance._predictions_df_val = instance._predictions_df[instance._predictions_df["set"] == "v1"]

display(instance._predictions_df_val.head())


BASELINE MODEL: PROBABILITIES FOR ALL IMAGES



lesion_id  num_images      image_id   dx  label dx_type   age   sex  \
0  HAM_0000118           2  ISIC_0027419  bkl      0   histo  80.0  male   
1  HAM_0000118           2  ISIC_0025030  bkl      0   histo  80.0  male   
2  HAM_0002730           2  ISIC_0026769  bkl      0   histo  80.0  male   
3  HAM_0002730           2  ISIC_0025661  bkl      0   histo  80.0  male   
4  HAM_0001466           2  ISIC_0031633  bkl      0   histo  75.0  male   

  localization set  prob_other  prob_mel  prob_akiec   prob_nv  prob_bcc  
0        scalp  ta    0.190499  0.589043    0.070307  0.127727  0.022423  
1        scalp  t1    0.325244  0.116826    0.034226  0.438181  0.085523  
2        scalp  va    0.025946  0.010540    0.299041  0.045798  0.618675  
3        scalp  v1    0.039586  0.041373    0.251411  0.027629  0.640001  
4          ear  va    0.110923  0.402103    0.215037  0.059775  0.212162


BASELINE MODEL: PROBABILITIES AND PREDICTIONS FOR ALL IMAGES



lesion_id  num_images      image_id   dx  label dx_type   age   sex  \
0  HAM_0000118           2  ISIC_0027419  bkl      0   histo  80.0  male   
1  HAM_0000118           2  ISIC_0025030  bkl      0   histo  80.0  male   
2  HAM_0002730           2  ISIC_0026769  bkl      0   histo  80.0  male   
3  HAM_0002730           2  ISIC_0025661  bkl      0   histo  80.0  male   
4  HAM_0001466           2  ISIC_0031633  bkl      0   histo  75.0  male   

  localization set  prob_other  prob_mel  prob_akiec   prob_nv  prob_bcc  pred  
0        scalp  ta    0.190499  0.589043    0.070307  0.127727  0.022423     2  
1        scalp  t1    0.325244  0.116826    0.034226  0.438181  0.085523     4  
2        scalp  va    0.025946  0.010540    0.299041  0.045798  0.618675     1  
3        scalp  v1    0.039586  0.041373    0.251411  0.027629  0.640001     1  
4          ear  va    0.110923  0.402103    0.215037  0.059775  0.212162     2


BASELINE MODEL: PROBABILITIES AND PREDICTIONS FOR IMAGES IN VALIDATION SET



lesion_id  num_images      image_id   dx  label dx_type   age     sex  \
3   HAM_0002730           2  ISIC_0025661  bkl      0   histo  80.0    male   
5   HAM_0001466           2  ISIC_0027850  bkl      0   histo  75.0    male   
7   HAM_0002761           2  ISIC_0029068  bkl      0   histo  60.0    male   
11  HAM_0004234           2  ISIC_0029396  bkl      0   histo  85.0  female   
13  HAM_0001949           2  ISIC_0025767  bkl      0   histo  70.0    male   

   localization set  prob_other  prob_mel  prob_akiec   prob_nv  prob_bcc  \
3         scalp  v1    0.039586  0.041373    0.251411  0.027629  0.640001   
5           ear  v1    0.118692  0.143967    0.113587  0.164686  0.459067   
7          face  v1    0.041351  0.002828    0.025426  0.131957  0.798437   
11        chest  v1    0.045510  0.151270    0.007371  0.789996  0.005853   
13        trunk  v1    0.712821  0.006769    0.000950  0.272685  0.006774   

    pred  
3      1  
5      1  
7      1  
11     4  
13     0

In [428]:
# from evaluation import weighted_average_f, confusion_matrix_with_metric

instance = resnet18mc_base
map_labels = instance.label_codes

A = instance._predictions_df_val['label']
B = instance._predictions_df_val['pred']
AxB = pd.crosstab(A,B,margins=True,dropna=False)

beta = 2

instance._cm = confusion_matrix_with_metric(AxB=AxB,
                                                lst=None,
                                                full_pad=True,
                                                func=weighted_average_f,
                                                beta=beta,
                                                weights=None,
                                                percentage=False,
                                                map_labels=map_labels)

print_header("BASELINE MODEL: confusion matrix (validation set)")

print(f"Overall evaluation metric at bottom right: average class-wise F{beta} score".upper())
display(instance._cm.fillna('_'))

C = instance._predictions_df['label']
D = instance._predictions_df['pred']
CxD = pd.crosstab(C,D,margins=True,dropna=False)

confusion_matrix_with_metric(AxB=CxD,
                                                lst=None,
                                                full_pad=True,
                                                func=weighted_average_f,
                                                beta=beta,
                                                weights=None,
                                                percentage=False,
                                                map_labels=map_labels)


BASELINE MODEL: CONFUSION MATRIX (VALIDATION SET)

OVERALL EVALUATION METRIC AT BOTTOM RIGHT: AVERAGE CLASS-WISE F2 SCORE


predicted    other       bcc       mel    akiec        nv    All    recall
actual                                                                    
other           78        19        46       11        71    225  0.346667
bcc              5        46         4        9        18     82  0.560976
mel              9         2        65        9        69    154  0.422078
akiec            7        20         6       20         4     57  0.350877
nv              35         9        43        5     1,259  1,351  0.931902
All            134        96       164       54     1,421  1,869         _
precision  0.58209  0.479167  0.396341  0.37037  0.885996      _   0.52265

predicted     other       bcc       mel     akiec        nv     All    recall
actual                                                                       
other           612        86       312        53       293   1,356  0.451327
bcc              14       367        26        21        86     514  0.714008
mel              45         7       587        36       438   1,113  0.527403
akiec            19        70        14       200        24     327  0.611621
nv              184        54       278        19     6,170   6,705  0.920209
All             874       584     1,217       329     7,011  10,015       NaN
precision  0.700229  0.628425  0.482334  0.607903  0.880046     NaN  0.644286